cmd/ltkeycloak/main.go

```diff
@@ -0,0 +1,468 @@
+	if err != nil {
+		return fmt.Errorf("failed to read load test configuration: %w", err)
+	}
+
+	// Guess where the deployment configuration is located
+	deploymentConfig, err := deployment.ReadConfig(filepath.Join(filepath.Dir(ltConfigPath), "deployer"+filepath.Ext(ltConfigPath)))
+	if err != nil {
+		return fmt.Errorf("failed to read deployment configuration: %w", err)
+	}
+
+	var keycloakHost string
+	keycloakHost, err = cmd.Flags().GetString("keycloak-host")
+	if err != nil {
+		return fmt.Errorf("failed to read flag: %w", err)
+	}
+
+	// Use the terraform output terraform host if a manual one is not provided. Useful for development.
+	if keycloakHost == "" {
+		t, err := terraform.New("", *deploymentConfig)
+		if err != nil {
+			return fmt.Errorf("failed to create terraform client: %w", err)
+		}
+		terraformOutput, err := t.Output()
+		if err != nil {
+			return fmt.Errorf("failed to get terraform output: %w", err)
+		}
+		if len(terraformOutput.KeycloakDatabaseCluster.Endpoints) == 0 {
+			return fmt.Errorf("keycloak database cluster not found in terraform output")
+		}
+		keycloakHost = terraformOutput.KeycloakDatabaseCluster.Endpoints[0]
+	}
+
+	dryRun, err := cmd.Flags().GetBool("dry-run")
+	if err != nil {
+		return fmt.Errorf("failed to read flag: %w", err)
+	}
+
+	mmClient := model.NewAPIv4Client(cfg.ConnectionConfiguration.ServerURL)
+
+	_, _, err = mmClient.Login(cmd.Context(), cfg.ConnectionConfiguration.AdminEmail, cfg.ConnectionConfiguration.AdminPassword)
+	if err != nil {
+		return fmt.Errorf("failed to login to mattermost: %w", err)
+	}
+
+	userPassword, err := cmd.Flags().GetString("set-user-password-to")
+	if err != nil {
+		return fmt.Errorf("failed to read flag: %w", err)
+	}
+
+	updateExistingUsers, err := cmd.Flags().GetBool("update-existing-users")
+	if err != nil {
+		return fmt.Errorf("failed to read flag: %w", err)
+	}
+
+	keycloakClient := gocloak.NewClient(keycloakHost)
+	ctx := context.Background()
+	token, err := keycloakClient.LoginAdmin(
+		ctx,
+		deploymentConfig.ExternalAuthProviderSettings.KeycloakAdminUser,
+		deploymentConfig.ExternalAuthProviderSettings.KeycloakAdminPassword,
+		"master", // TODO: Allow specifying the master realm
+	)
+	if err != nil {
+		return fmt.Errorf("failed to login to keycloak: %w", err)
+	}
+
+	if !dryRun {
+		// Create group for migrated users if it does not exist
+		_, err := keycloakClient.CreateGroup(ctx, token.AccessToken, deploymentConfig.ExternalAuthProviderSettings.KeycloakRealmName, gocloak.Group{
+			Name: gocloak.StringP(keycloakMigratedGroup),
+		})
+		if err != nil {
+			// Ignore if group already exists
+			if apiErr, ok := err.(*gocloak.APIError); ok && apiErr.Code != 409 {
+				return fmt.Errorf("failed to create group for user migrations in keycloak: %w", err)
+			}
+		}
+	}
+
+	start := 0
+	perPage := 100
+	for {
+		requestCtx, cancel := context.WithTimeout(context.Background(), operationTimeout)
+		defer cancel()
```

Comment:
Note that since this is a loop, cancel might not get called until the entire function exits. Not a big deal, but worth keeping in mind. Ideally, we call `cancel` as soon as the operation is done.